File: python/day14.py
import re

def parse(rs):
    m = re.match("(\w+) can fly (\d+) km/s for (\d+) seconds, but then must rest for (\d+) seconds.", rs)
    name = m.group(1)
    speed = int(m.group(2))
    time = int(m.group(3))
    rest = int(m.group(4))
    return (name, speed, time, rest)

def simulate(reindeers, limit):
    states = []
    for r in reindeers:
        # name, state, time left in state, distance, points
        states.append([r[0], "RUNNING", r[2], 0, 0])
    reindeers = {r[0]: r[1:] for r in reindeers}

    for _ in range(limit):
        for s in states:
            if s[1] == "RUNNING":
                s[3] += reindeers[s[0]][0]
                s[2] -= 1
                if s[2] == 0:
                    s[2] = reindeers[s[0]][2]
                    s[1] = "RESTING"
            else:
                s[2] -= 1
                if s[2] == 0:
                    s[2] = reindeers[s[0]][1]
                    s[1] = "RUNNING"
        m = max(s[3] for s in states)
        for s in states:
            if s[3] == m:
                s[4] += 1

    states.sort(key=lambda x: x[3], reverse=True)
    distance = (states[0][0], states[0][3])
    states.sort(key=lambda x: x[4], reverse=True)
    points = (states[0][0], states[0][4])
    return distance, points

with open("day14.in") as fin:
    contents = fin.read().strip().split("\n")

reindeers = [parse(c) for c in contents]
distance, points= simulate(reindeers, 2503)
print(distance)
print(points)

File: aoc2015/src/bin/day14.rs
extern crate fancy_regex;

use std::fs::File;
use std::io::BufReader;
use std::io::BufRead;
use std::cmp::max;
use std::str::FromStr;
use fancy_regex::Regex;

enum State {
    Flying(u32),
    Resting(u32),
}

struct Reindeer {
    speed: u32,
    flight_time: u32,
    rest_time: u32,
}

struct Racer {
    state: State,
    distance: u32,
    points: u32,
}

impl FromStr for Reindeer {
    type Err = Box<dyn std::error::Error>;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let re = Regex::new(r"\w+ can fly (\d+) km/s for (\d+) seconds, but then must rest for (\d+) seconds.")?;
        let capture = re.captures(s)?.unwrap();

        let speed = capture.get(1).unwrap().as_str().parse()?;
        let flight_time = capture.get(2).unwrap().as_str().parse()?;
        let rest_time = capture.get(3).unwrap().as_str().parse()?;

        let r = Reindeer{
            speed: speed,
            flight_time: flight_time,
            rest_time: rest_time,
        };

        Ok(r)
    }
}

fn simulate(reindeers: &[Reindeer], limit: usize) -> (u32, u32) {
    let mut racers: Vec<Racer> = reindeers.iter().map(|r| Racer {state: State::Flying(r.flight_time), distance: 0, points: 0}).collect();

    for _ in 0..limit {
        let mut m = 0;
        for (i, r) in racers.iter_mut().enumerate() {
            match r.state {
                State::Flying(f) => {
                    r.distance += reindeers[i].speed;
                    r.state = if f == 1 { State::Resting(reindeers[i].rest_time) } else { State::Flying(f-1) };
                }
                State::Resting(x) => {
                    r.state = if x == 1 { State::Flying(reindeers[i].flight_time) } else { State::Resting(x-1) };
                }
            }
            m = max(m, r.distance);
        }
        for r in racers.iter_mut() {
            if r.distance == m {
                r.points += 1;
            }
        }
    }

    let mut max_distance = 0;
    let mut max_points = 0;
    for r in racers.iter() {
        max_distance = max(max_distance, r.distance);
        max_points = max(max_points, r.points);
    }
    
    (max_distance, max_points)
}


fn main() {
    let f = File::open("inputs/day14.in").unwrap();
    let file = BufReader::new(&f);

    let reindeers: Vec<Reindeer> = file.lines().map(|l| l.unwrap().parse().unwrap()).collect();
    let (part1, part2) = simulate(&reindeers, 2503);

    println!("Part 1: {}", part1);
    println!("Part 2: {}", part2);
}
